### PR TITLE
Add Sphinx docs customization for multimethod

### LINKED ISSUE #1087: Show multimethod signatures in HTML docs
Currently the HTML docs do not show all of the signatures of methods implemented with `multimethod`.  Also see #1049.

### PR DESCRIPTION
**Pull request by lorenzncode**
This adds customization to better handle multimethod in the HTML docs.  To resolve #1087.

Now multimethod is handled in a similar way as `overload` and `singledispatch`.

### CODE REVIEW

**Comment by lorenzncode** [comment]
Here is an example of the customized autosummary output (API Reference):
image

It shows ellipsis in place of the input parameters now.

This provides a hint, and following the link, multiple signatures are displayed.  
image


The Parameters field list continues to show a single set of parameters for one of the signatures similar to overloads.  The source link is available to show the implementation.

**Comment by jmwright** [comment]
@lorenzncode I think this looks good. Is it ready for review?

**Comment by lorenzncode** [comment]
Yes, this is ready for review. Thanks for taking a look.

**Comment by Peque** [comment]
@jmwright Not the time, but thanks for the ping! :blush:

**Comment by lorenzncode** [comment]
Thanks @jmwright.  Yes I reproduced the warning. I've pushed a change to address the warning by setting the `override=True` flag.  Since this is an intentional override, sphinx will no longer issue the warning.

**Comment by jmwright** [comment]
Thanks @lorenzncode 

Other than the two other comments I think this PR looks good. I'm not sure who else will review it. @adam-urbanczyk are you available to do it?

**Comment by dcowden** [comment]
@jmwright i can review, I probably can get to it early next week

**Comment by jmwright** [comment]
@dcowden Great thanks!

**Review by adam-urbanczyk — APPROVED** [review]
LGTM, would it be possible (and not too much work) to generate three parameter lists under each function prototype?

afbeelding

**Comment by dcowden** [comment]
+1

**Review by dcowden — APPROVED** [review]
Definitely not a sphinx expert here, but looks ok to me

**Review by jmwright — APPROVED** [review]
@lorenzncode Thanks! Feel free to merge if you think it's ready. You might want to start a new issue for the comment made by @adam-urbanczyk . I think this is a nice stand-alone PR as-is, but that's a possible future change.

File: doc/ext/sphinx_autodoc_multimethod.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+                continue
+
+            self.bridge.result = StringList()  # initialize for each documenter
+            full_name = real_name
+            if not isinstance(obj, ModuleType):
+                # give explicitly separated module name, so that members
+                # of inner classes can be documented
+                full_name = modname + "::" + full_name[len(modname) + 1 :]
+            # NB. using full_name here is important, since Documenters
+            #     handle module prefixes slightly differently
```
[the remark this answers is not in the document]
Comment (lorenzncode) [reply doc/ext/sphinx_autodoc_multimethod.py:138]: This extension creates a subclass of `sphinx.ext.autodoc.MethodDocumenter` overriding method `format_signature` and a sublcass of `sphinx.ext.autosummary.AutoSummary` overriding `get_items`.  Actually this comment comes from the original implementation of `get_items`.  I didn't change the formatting other than run black.  I would suggest to keep the extra spaces as is to help diff in the future say when checking for compatibility with a new Sphinx release (I'll add a note on maintenance).

File: cadquery/sketch.py [hunk 2]
```diff
@@ -782,7 +782,7 @@ def arc(
         forConstruction: bool = False,
     ) -> T:
         """
-        Construct an arc starting at p1, through p2, ending at p3
```
[the remark this answers is not in the document]
Comment (lorenzncode) [reply cadquery/sketch.py:785]: Right, this was intentionally reverted back to the original docstring.  The original docstring has a generic description that is applicable to all of the dispatch methods which is better than a description specific to only one of the methods.

**Comment by lorenzncode** [comment]
Notes on maintenance:

When updating to a newer Sphinx version, test that the extension is working.  The extension depends on Sphinx autodoc and autosummary and may require updates depending on changes to Sphinx (there were none from v4.5 to v5.0 so far).

Compare `MultimethodAutosummary.get_items`, and `MethodDocumenter.format_signature` with the Sphinx implementation. 
The modifications are commented with `# -- multimethod customization`, `# -- end customization`.

**Comment by lorenzncode** [comment]
I pushed a minor change to a .gitignore entry added in this PR.  If there are no other issues, this is ready to squash and merge.

**Comment by jmwright** [comment]
Thanks @lorenzncode 

I re-ran the randomly falling test on Windows and it passed the second time.

**Comment by adam-urbanczyk** [comment]
> I agree the vertical field list for each prototype would be nice to try. Unfortunately it does not look easy to me.
> 
> This PR attempts to bring multimethod support more in line with how Sphinx handles `overload` and `singledispatchmethod` where Sphinx supports a list of signatures. This was mostly a matter of identifying multimethod usage and appending to the signature list. Sphinx does not show the Parameter field list for each signature today for say `singledispatchmethod` so I think more significant customization would be required.

No worries, thanks for the hard work!

**Pull request merged**